doc/2/plugins/guides/hooks/index.md

```diff
@@ -13,7 +13,189 @@ Hooks can only listen: the received information cannot be changed. And Kuzzle do
+  'event-name': {
+    callback: <fn>,
+    withRequest: {
+      // Koncorde filter
+    },
+    withDocument: {
+      // Koncorde filter
+    },
+  }
+};
+```
+
+## Unconditional listening
+
+If no filter is set, then the callback is called each time the event is triggered.  
+The callback arguments depend on the selected event. 
+
+```js
+this.hooks = {
+  'collection:afterList': {
+    callback: request => {
+      // this callback will be called each time the event is triggered
+    }
+  },
+  'document:generic:beforeWrite': {
+    callback: (documents, request) => {
+      // this callback will be called each time the event is triggered
+    }
+  }
+};
+```
+
+## Filter on Request
+
+If a filter is set in the `withRequest` property, then the callback will be executed only if the filter match the request.  
+
+This allow you to easily decide when you want to execute a precise callback:
+ - for particular index and/or collection
+ - for a specific user
+ - when the request body contain a certain field
+ - etc
+
+```js
+this.hooks = {
+  'collection:afterList': {
+    callback: request => {
+      // this callback will be called only 
+      // when request.input.resource.index is equals to 'nyc-open-data'
+    },
+    withRequest: {
+      equals: { 'input.resource.index': 'nyc-open-data' }
+    }
+  }
+}
+```
+
+::: warning
+The `withRequest` filter is only available for events that receive a Request object, with other events Kuzzle will throw an error and interrupt its initialization sequence. (eg: [core:overload](/core/2/plugins/guides/events/core:overload))  
+:::
+
+```js
+this.hooks = {
+  'core:overload': {
+    callback: fill => {
+      // this event does not have a request object in its arguments 
+      // so Kuzzle with throw an error and shutdown
+    },
+    withRequest: {
+      equals: { 'input.resource.index': 'nyc-open-data' }
+    }
+  }
+}
+```
+
+## Filter on Document
+
+This filter is meant to be used with [Generic Document events](/core/2/plugins/guides/events/generic-document-events).  
+
+If a filter is set in the `withDocument` property, then the callback will be executed only with the documents matching that filter.  
+
+### Callback behavior
+
+The callback is called with two arguments: 
+ - `document`: one of the document matching the provided filter
+ - `request`: the request object
+
+The callback can return either the document or a promise resolving to it.  
+If `null` is returned, then the document will be excluded from further processing.  
+
+```js
+this.hooks = {
+  'generic:document:beforeWrite': {
+    callback: (document, request) => {
+      // This callback will be executed only with document 
+      // containing an "age" property
+      // 
+      // Then if this age is less than 18, the document is filtered
+      if (document.body.age >= 18) {
+        return document
+      }
+      else {
+        return null;
+      }
+    },
+    withDocument: {
+      exists: 'age'
+    }
+  }
+}
+```
```

Comment:
Why don't we use a [range filter](https://docs.kuzzle.io/core/2/guides/cookbooks/realtime-api/terms/#range) instead?

```suggestion
this.hooks = {
  'generic:document:beforeWrite': {
    callback: (document, request) => {
      // This callback will be executed only on documents 
      // containing an "age" property greater or equal than 18
    },
    withDocument: {
      range: { age: { gte: 18 } }
    }
  }
}
```
```

Reply:
With the range filter, the callback will be called only for document with `age` greater than 18, and the callback will not be called document with `age` less than 18 so we cannot filter them.  
The filter allow to call the callback only with matching document, he is not actually filtering the documents

Reply:
In any case, your callback is wrong: the received payload from that event is an array of documents.

Reply:
Yes but my point is that the callback is unitary called with each matching document and not with a filtered list of documents

Reply:
I didn't get that. And I don't like it, because this means that large queries such as mCreate and the like, cannot be handled by plugins unless they process documents 1 by 1.

Reply:
I agree but in this case we cannot use this pipe to continue the processing only with a subset of the original documents.  

For example, I want my action to be executed only on document where `age > 18` but that didn't mean that I don't want the document with `age <= 18` to be excluded of the rest of the processing.

If you want to process the array of documents, you can only provide the `withRequest` filter and you will have to manually filter documents that you don't want.

Reply:
First, there is another thing that I don't like, and it is that your solution changes the callback payload. Usually, that event sends an array of documents, but if filtered, it receives a single document object instead. This change is made silently, and it isn't obvious in the documentation (hell, I didn't get it either with the documentation under my nose).

And as I said, the solution you propose makes that feature unusable with `m*` API routes, because you lose the ability to perform operations on whole sets of documents, with far better performances than performing these operations times the number of documents.

The solution I proposed initially was a hard filter only for the callback with the filters, if you read it carefully, you'll see that the original documents array is mutated with the result of the pipe function, to report all changes to their corresponding, original array indexes. Before invoking the next pipe.

To me, there are 2 ways of handling these filters more or less gracefully:
1. filter the documents sent to the callback according to the provided Koncorde filter. This is the solution I proposed, and basically, this means 1/ building a subarray containing only the matching documents, 2/ provide this subarray to the pipe, 3/ mutate the original array with the pipe's results, and 4/ continue with the next pipe function
1. do not alter the documents array in any way. Instead, make the Koncorde filter act as a filter on the whole array: if at least one document matches, then the pipe function is executed with the whole set of documents

Solution 1 is the nicest for plugin developers.
Solution 2 is the safest and the most maintenable for us.